omod/src/main/java/org/openmrs/module/owa/web/controller/OwaRestController.java

```diff
@@ -153,66 +128,37 @@
-				"File name must be passed for installation"
-			);
-		}
-		URL downloadUrl = null;
-		if (ResourceUtils.isUrl(url)) {
-			downloadUrl = new URL(url);
-		}
-		if (Context.hasPrivilege("Manage OWA")) {
-			String message;
+		if(Context.hasPrivilege("Manage OWA")){
 			HttpSession session = request.getSession();
-			if (!url.isEmpty()) {
-				InputStream inputStream = ModuleUtil.getURLStream(downloadUrl);
-				log.info("Url pathname: " + downloadUrl.getPath());
-				fileName += ".zip";
-				final String fileMatch=  fileName;
-				File file = ModuleUtil.insertModuleFile(inputStream, fileName);
-				try (ZipFile zip = new ZipFile(file)) {
-					if (zip.size() == 0) {
-						message = messageSourceService.getMessage("owa.blank_zip");
-						file.delete();
-						log.warn("Zip file is empty");
-						session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, message);
-						response.sendError(500, message);
-					} else {
-						ZipEntry entry = zip.getEntry("manifest.webapp");
-						if (entry == null) {
-							message = messageSourceService.getMessage("owa.manifest_not_found");
-							log.warn("Manifest file could not be found in app");
-							file.delete();
-							session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, message);
-							response.sendError(500, message);
-						} else {
-							String contextPath = request.getScheme() + "://" + request.getServerName() + ":"
-									+ request.getServerPort() + request.getContextPath();
-							appManager.installApp(file, fileName, contextPath);
-							file.delete();
-							message = messageSourceService.getMessage("owa.app_installed");
-							response.setStatus(200);
-							session.setAttribute(WebConstants.OPENMRS_MSG_ATTR, message);
-						}
-					}
-				}
-				catch (Exception e) {
-					message = messageSourceService.getMessage("owa.not_a_zip");
-					log.warn("App is not a zip archive");
-					file.delete();
-					session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, message);
-					response.sendError(500, message);
-				}
-			} else {
+			URL downloadUrl = null;
+			String installUrl = urlObject.getUrlValue();
+			if(installUrl == null || installUrl.isEmpty()){
+				message = messageSourceService.getMessage("owa.invalid_url");
+				log.warn("Invalid url");
+				session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, message);
+				response.sendError(400, message);
+				return appList;
+			}
+			if (ResourceUtils.isUrl(installUrl)) {
+				downloadUrl = new URL(installUrl);
+			}
+			fileName = this.getPassedFileName(installUrl, downloadUrl);
+			if(fileName == null || fileName.isEmpty()){
 				message = messageSourceService.getMessage("owa.invalid_url");
 				log.warn("Invalid url");
 				session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, message);
 				response.sendError(400, message);
+				return appList;
 			}
+
+			InputStream inputStream = ModuleUtil.getURLStream(downloadUrl);
+			log.info("Url pathname: " + downloadUrl.getPath());
+			fileName += ".zip";
```

Comment:
`this.getPassedFileName(installUrl, downloadUrl)` should return the full fileName(including the `.zip`). So we don't manually add the `.zip` back like this.

Reply:
Have modified the getPassedFileName function to return the file name with the `.zip` attached